File: civitasjava/src/Civitas/MazoSorpresas.java
package civitas;

import java.util.ArrayList;
import java.util.Collections;

public class MazoSorpresas {
    private ArrayList<Sorpresa> sorpresas;  //Almacena las cartas Sorpresa
    private boolean barajada;               //Indica si se ha barajado
    private int usadas;                     //Numero de cartas ya usadas
    private boolean debug; 
    private ArrayList<Sorpresa> cartasEspeciales;//Almacena la carta SALIRCARCEL
    private Sorpresa ultimaSorpresa;        //Ultima sorpresa que ha salido
    
    /**
     * Inicializa sorpresas,cartasEspeciales,barajada y usadas
     */
    private void init(){
        barajada = false;
        usadas = 0;
        cartasEspeciales = new ArrayList<>();
        sorpresas = new ArrayList<>();
    }
    /**
     * Constructor sin parametros
     */
    MazoSorpresas(){
        init();
        debug = false;
    }
    /**
     * Constructor por parametros
     * @param d 
     */
    MazoSorpresas(boolean d){
        debug = d;
        init();
        if(debug){
            Diario.getInstance().ocurreEvento("Debug true");
        }
    }
    
    /**
     * Añade la Sorpresa del parametro al final
     * @param s : sorprese que se quiere añadir
     */
    void alMazo(Sorpresa s){
        if(barajada == false){
            sorpresas.add(s);
        }
    }
    
    /**
     * Si cumple las condiciones se baraja el mazo, la primera sorpresa pasa a 
     * ser la ultima y se devuelve por referencia
     * @return 
     */
    Sorpresa siguiente(){
        if(barajada == false || usadas == sorpresas.size() && debug == false){
            Collections.shuffle(sorpresas);
            usadas = 0;
            barajada = true;
        }
                
        usadas++;

        Sorpresa aux = sorpresas.get(0);
        sorpresas.add(sorpresas.size()-1, aux);
        sorpresas.remove(0);

        ultimaSorpresa = aux;
        
        return ultimaSorpresa;
    }
    
    /**
     * Si sorpresa es una carta especial y esta en el mazo, se 
     * quita y se añade a cartasEspeciales
     * @param sorpresa : sorpresa que se quiere inhabilitar
     */
    void inhabilitarCartaEspecial(Sorpresa sorpresa){
        Sorpresa aux;

        for(int i=0; i<sorpresas.size(); i++){
            if(sorpresas.get(i) == sorpresa){

                aux = sorpresas.get(i);
                sorpresas.remove(i);
                cartasEspeciales.add(aux);
                Diario.getInstance().ocurreEvento("Una carta movida de "
                        + "sorpresa a especiales");
            }
        }
    }
    
    /**
     * Si sorpresa esta en cartasEspeciales, se saca y se añade a la coleccion
     * de sorpresas para habilitarla
     * @param sorpresa : sorpresa que se quiere habilitar
     */
    void habilitarCartaEspecial(Sorpresa sorpresa){
        Sorpresa aux;

        for(int i=0; i<cartasEspeciales.size(); i++){
            if(cartasEspeciales.get(i) == sorpresa){
                aux = cartasEspeciales.get(i);
                cartasEspeciales.remove(i);
                sorpresas.add(aux);
                Diario.getInstance().ocurreEvento("Una carta movida de "
                        + "especiales a sorpresa");
            }
        }
    }
	
    Sorpresa getUltimaSorpresa(){
        return ultimaSorpresa;
    }

    Sorpresa getSorpresas(int i){
        return sorpresas.get(i);
    }

    int getTamSorpresas(){
        return sorpresas.size();
    }
    
}
    
    

File: civitasruby/lib/mazo_sorpresas.rb
# encoding:utf-8
module Civitas
class Mazo_sorpresas
   
  def init
    @sorpresas=Array.new
    @cartasEspeciales=Array.new
    @barajada=false
    @usadas=0
    @ultimaSorpresa=nil
  end
  
  def initialize(d=nil)
    if(d==nil)
    init 
    @debug=false
    else
      @debug=d
     init
      if(d)
        Diario.instance.ocurre_evento('Debug ON(mazo no se baraja)')
      end
    end  
  end
  
  def al_mazo(s)
    if(!@barajada)
      @sorpresas<<s  
    end
  end
  
  def siguiente 
    if((!@barajada)||(@usadas == @sorpresas.size)&&(@debug = false))
      @sorpresas.shuffle
      @usadas = 0
      @barajada = true
    end
    @usadas = @usadas + 1
    @ultimaSorpresa = @sorpresas[0]
    @sorpresas.delete_at(0)
    @sorpresas<<@ultimaSorpresa
    return @ultimaSorpresa
  end
  
  
  def inhabilitar_carta_especial(sorpresa)
    for i in 0...@sorpresas.size
      if(@sorpresas[i]==sorpresa)
          a=@sorpresas[i]
          @sorpresas.delete_at(i)
          @cartasEspeciales<<a
          Diario.instance.ocurre_evento('Se ha inhabilitado una carta del mazo sorpresas y se ha añadido al de cartas especiales')
      end
    end
  end
  
  def habilitar_carta_especial(sorpresa)
    for i in 0...@cartasEspeciales.size
      if(@cartasEspeciales[i]==sorpresa)
          a=@cartasEspeciales[i]
          @cartasEspeciales.delete_at(i)
          @sorpresas<<a
          Diario.instance.ocurre_evento('Se ha habilitado una carta al mazo sorpresas proveniente de las cartas especiales')
      end
    end  
  end
  
  
 
  private :init
  
  end
  
end